buildSrc/src/main/kotlin/datadog/gradle/plugin/naming/InstrumentationNamingPlugin.kt

```diff
@@ -0,0 +1,180 @@
+            appendLine()
+            appendLine("To exclude specific modules or customize suffixes, configure the plugin:")
+            appendLine("  instrumentationNaming {")
+            appendLine("    exclusions.set(listOf(\"module-name\"))")
+            appendLine("    suffixes.set(listOf(\"-common\", \"-stubs\"))")
+            appendLine("  }")
+          }
+          throw GradleException(errorMessage)
+        } else {
+          target.logger.lifecycle("✓ All instrumentation modules follow naming conventions")
+        }
+      }
+    }
+  }
+
+  private fun validateInstrumentations(
+    instrumentationsDir: File,
+    exclusions: Set<String>,
+    suffixes: List<String>
+  ): List<NamingViolation> {
+    val violations = mutableListOf<NamingViolation>()
+
+    fun hasBuildFile(dir: File): Boolean = dir.listFiles()?.any {
+      it.name == "build.gradle" || it.name == "build.gradle.kts"
+    } ?: false
+
+    fun traverseModules(currentDir: File, parentName: String?) {
+      currentDir.listFiles { file -> file.isDirectory }?.forEach childLoop@{ childDir ->
+        val moduleName = childDir.name
+
+        // Skip build directories and other non-instrumentation directories
+        if (moduleName in setOf("build", "src", ".gradle")) {
+          return@childLoop
+        }
+
+        val childHasBuildFile = hasBuildFile(childDir)
+        val nestedModules = childDir.listFiles { file -> file.isDirectory }?.filter { hasBuildFile(it) } ?: emptyList()
+
+        if (childHasBuildFile && moduleName !in exclusions) {
+          val relativePath = childDir.relativeTo(instrumentationsDir).path
+          if (parentName == null && nestedModules.isEmpty()) {
+            validateLeafModuleName(moduleName, relativePath, suffixes)?.let { violations.add(it) }
+          } else if (parentName != null) {
+            validateModuleName(moduleName, parentName, relativePath, suffixes)?.let { violations.add(it) }
+          }
+        }
+
+        // Continue traversing to validate deeply nested modules
+        if (nestedModules.isNotEmpty() || !childHasBuildFile) {
+          traverseModules(childDir, moduleName)
+        }
+      }
+    }
+
+    traverseModules(instrumentationsDir, null)
+
+    return violations
+  }
+
+  private fun validateModuleName(
+    moduleName: String,
+    parentName: String,
+    relativePath: String,
+    suffixes: List<String>
+  ): NamingViolation? {
+    // Rule 1: Module name must end with version pattern or one of the configured suffixes
+    validateVersionOrSuffix(moduleName, relativePath, suffixes)?.let { return it }
+
+    // Rule 2: Module name must contain parent directory name
+    if (!moduleName.contains(parentName, ignoreCase = true)) {
+      return NamingViolation(
+        relativePath,
+        "Module name '$moduleName' should contain parent directory name '$parentName'"
+      )
+    }
+
+    return null
+  }
```

Comment:
How about to use Kotlin null safety tick here like this:
```suggestion
  private fun validateModuleName(
    moduleName: String,
    parentName: String,
    relativePath: String,
    suffixes: List<String>
  ): List<NamingViolation> {
    // Rule 1: Module name must end with version pattern or one of the configured suffixes
    validateVersionOrSuffix(moduleName, relativePath, suffixes)?.let { return it }

    // Rule 2: Module name must contain parent directory name
    if (!moduleName.contains(parentName, ignoreCase = true)) {
      return listOf(NamingViolation(
        relativePath,
        "Module name '$moduleName' should contain parent directory name '$parentName'"
      ))
    }

    return emptyList()
  }
```

than in place where it will be used it will be something like:
`violations.addAll(validateModuleName(moduleName, parentName, relativePath, suffixes))`